crates/vm/src/arch/interpreter.rs

```diff
@@ -62,17 +72,46 @@ struct TerminatePreCompute {
     exit_code: u32,
 }
 
-macro_rules! execute_with_metrics {
-    ($span:literal, $pc_base:expr, $exec_state:expr, $pre_compute_insts:expr) => {{
+macro_rules! run {
+    ($span:literal, $interpreter:ident, $exec_state:ident, $ctx:ident) => {{
         #[cfg(feature = "metrics")]
         let start = std::time::Instant::now();
         #[cfg(feature = "metrics")]
         let start_instret = $exec_state.instret;
 
-        // SAFETY: pre_compute_insts contains valid function pointers and pre-computed data
-        info_span!($span).in_scope(|| unsafe {
-            execute_trampoline($pc_base, $exec_state, $pre_compute_insts);
-        });
+        info_span!($span).in_scope(|| -> Result<(), ExecutionError> {
+            // SAFETY:
+            // - it is the responsibility of each Executor to ensure that pre_compute_insts contains
+            //   valid function pointers and pre-computed data
+            #[cfg(not(feature = "tco"))]
+            unsafe {
+                tracing::debug!("execute_trampoline");
+                execute_trampoline(&mut $exec_state, &$interpreter.pre_compute_insns);
+            }
+            #[cfg(feature = "tco")]
+            {
+                tracing::debug!("execute_tco");
+                let handler = $interpreter
+                    .get_handler($exec_state.pc)
+                    .ok_or(ExecutionError::PcOutOfBounds($exec_state.pc))?;
```

Comment:
set `exit_code`?

Reply:
the `execute_from_state` fn only uses `exit_code` internally, so here I just want the entire function to return error